drivers/i2c/i2c_gd32.c

```diff
@@ -0,0 +1,684 @@
+	data->current->buf++;
+
+	if (data->xfer_len && (data->current->len == 0U)) {
+		data->current++;
+	}
+}
+
+static void i2c_gd32_handle_rbne(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->xfer_len) {
+		switch (data->xfer_len) {
+		case 1:
+			/* Reception for N == 1. */
+			data->xfer_len--;
+			i2c_gd32_xfer_read(data, cfg);
+
+			k_sem_give(&data->sync_sem);
+
+			break;
+		case 2:
+			__fallthrough;
+		case 3:
+			/*
+			 * Reception for N == 2 or
+			 * reception for N > 3 when xfer_len == 3.
+			 */
+			I2C_CTL1(cfg->reg) &= ~I2C_CTL1_BUFIE;
+			break;
+		default:
+			/* Reception for N > 3 when xfer_len > 3. */
+			data->xfer_len--;
+			i2c_gd32_xfer_read(data, cfg);
+			break;
+		}
+	} else {
+		if ((data->current->flags & I2C_MSG_STOP) &&
+		    (data->is_stopped == 0U)) {
+			data->is_stopped = 1U;
+			I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+		}
+
+		k_sem_give(&data->sync_sem);
+	}
+}
+
+static void i2c_gd32_handle_tbe(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->xfer_len) {
+		data->xfer_len--;
+		if (data->xfer_len == 0U) {
+			/*
+			 * This is the last data to transmit, disable the TBE irq.
+			 * Use the BTC irq to indicate the write data complete state.
+			 */
+			I2C_CTL1(cfg->reg) &= ~I2C_CTL1_BUFIE;
+		}
+		i2c_gd32_xfer_write(data, cfg);
+
+	} else {
+		/* There have some duplicate irq, use is_stopped to ignore it. */
+		if ((data->current->flags & I2C_MSG_STOP) &&
+		    (data->is_stopped == 0U)) {
+			data->is_stopped = 1U;
+			I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+		}
+
+		if (I2C_STAT0(cfg->reg) & I2C_STAT0_BTC) {
+			I2C_DATA(cfg->reg);
+		}
+
+		k_sem_give(&data->sync_sem);
+	}
+}
+
+static void i2c_gd32_handle_btc(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->current->flags & I2C_MSG_READ) {
+		uint32_t counter = 0U;
+
+		switch (data->xfer_len) {
+		case 2:
+			/*
+			 * Stop condition must be generated before reading the
+			 * last two bytes.
+			 */
+			if ((data->current->flags & I2C_MSG_STOP) &&
+			    (data->is_stopped == 0U)) {
+				data->is_stopped = 1U;
+				I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+			}
+
+			for (counter = 2U; counter > 0; counter--) {
+				data->xfer_len--;
+				i2c_gd32_xfer_read(data, cfg);
+			}
+
+			k_sem_give(&data->sync_sem);
+
+			break;
+		case 3:
+			I2C_CTL0(cfg->reg) &= ~I2C_CTL0_ACKEN;
+
+			data->xfer_len--;
+			i2c_gd32_xfer_read(data, cfg);
+
+			break;
+		default:
+			i2c_gd32_handle_rbne(dev);
+			break;
+		}
+	} else {
+		i2c_gd32_handle_tbe(dev);
+	}
+}
+
+static void i2c_gd32_handle_addsend(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	if (data->is_restart == 1U) {
+		I2C_STAT1(cfg->reg);
```

Comment:
The user guide suggests both stat0 and stat1 need to be read each time, is there a particular reason why you don't read the register as the first step here?

The I2C_STAT1() macro by itself is also a little confusing to look at, its not clear why its being done as at first glance it appears to do nothing at all, its only in looking at the data sheet and the hal did I understand the volatile read here which makes this, in my opinion, harder to understand.

Reply:
I2C_STAT0 already read at i2c_gd32_event_isr. Looks there need a comment explanation it.

Reply:
I did see that, and that makes sense, which is why the I2C_STAT1() macro without assignment or mutation looks even stranger at first glance. A comment would be helpful.